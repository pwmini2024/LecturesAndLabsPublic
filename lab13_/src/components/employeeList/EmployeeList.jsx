import React, { useState, useEffect } from "react";
import Loader from "../utils/Loader";

function EmployeeList() {
  const [employees, setEmployees] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetch("http://localhost:3001/employees")
      .then((response) => {
        if (!response.ok) {
          throw new Error("Failed to fetch");
        }
        return response.json();
      })
      .then((data) => {
        setEmployees(data);
        setLoading(false);
      })
      .catch((err) => {
        setError(err.message);
        setLoading(false);
      });
  }, []);

  if (loading) return <Loader />;
  if (error) return <div>Error: {error}</div>;

  return (
    <div className="employee-list">
      <h2>Employees</h2>
      <div className="employees">
        {employees.map((employee) => (
          <div key={employee.id} className="employee-card">
            <h3>{employee.name}</h3>
            <p>Age: {employee.age}</p>
            <p>Company: {employee.company}</p>
            <p>Email: {employee.email}</p>
            <p>Status: {employee.isActive ? "Active" : "Inactive"}</p>
          </div>
        ))}
      </div>
    </div>
  );
}

export default EmployeeList;
