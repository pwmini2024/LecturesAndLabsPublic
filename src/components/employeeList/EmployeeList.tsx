import React from 'react';

const EmployeeList: React.FC = () => {
    return (
        <></>
    );
}
export default EmployeeList;
