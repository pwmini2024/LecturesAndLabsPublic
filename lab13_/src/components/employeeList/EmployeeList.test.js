import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import EmployeeList from "./EmployeeList";

// Mock fetch globally
const mockFetch = jest.fn();

describe("EmployeeList Component", () => {
  // Setup and teardown
  beforeEach(() => {
    // Clear mock and set up default mock implementation
    mockFetch.mockClear();
    global.fetch = mockFetch.mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve([]),
      })
    );
  });

  // Cleanup after tests
  afterEach(() => {
    global.fetch = undefined;
  });

  // Integration test - loading state
  it("renders loading state initially", async () => {
    await act(async () => {
      render(<EmployeeList />);
    });
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  // Integration test - successful data fetching
  it("renders employees after successful fetch", async () => {
    const mockEmployees = [
      {
        id: "1",
        name: "John Doe",
        age: 30,
        company: "Test Corp",
        email: "john@test.com",
        isActive: true,
      },
    ];

    mockFetch.mockImplementationOnce(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockEmployees),
      })
    );

    await act(async () => {
      render(<EmployeeList />);
    });

    // Test loading state
    expect(screen.getByText("Loading...")).toBeInTheDocument();

    // Test loaded content
    await act(async () => {
      await waitFor(() => {
        expect(screen.getByText("John Doe")).toBeInTheDocument();
      });
    });

    // Test data display
    expect(screen.getByText("Age: 30")).toBeInTheDocument();
    expect(screen.getByText("Company: Test Corp")).toBeInTheDocument();
  });

  // Error handling test
  it("renders error message on fetch failure", async () => {
    mockFetch.mockImplementationOnce(() =>
      Promise.reject(new Error("Network error"))
    );

    await act(async () => {
      render(<EmployeeList />);
    });

    await act(async () => {
      await waitFor(() => {
        expect(screen.getByText(/Error:/)).toBeInTheDocument();
      });
    });
  });

  // Snapshot test
  it("matches snapshot when data is loaded", async () => {
    const mockEmployees = [
      {
        id: "1",
        name: "John Doe",
        age: 30,
        company: "Test Corp",
        email: "john@test.com",
        isActive: true,
      },
    ];

    mockFetch.mockImplementationOnce(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockEmployees),
      })
    );

    let container;
    await act(async () => {
      const rendered = render(<EmployeeList />);
      container = rendered.container;
    });

    await act(async () => {
      await waitFor(() => {
        expect(screen.getByText("John Doe")).toBeInTheDocument();
      });
    });

    expect(container).toMatchSnapshot();
  });
});
