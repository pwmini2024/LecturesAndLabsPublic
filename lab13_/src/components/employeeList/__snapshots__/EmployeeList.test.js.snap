// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeList Component matches snapshot when data is loaded 1`] = `
<div>
  <div
    class="employee-list"
  >
    <h2>
      Employees
    </h2>
    <div
      class="employees"
    >
      <div
        class="employee-card"
      >
        <h3>
          John Doe
        </h3>
        <p>
          Age: 
          30
        </p>
        <p>
          Company: 
          Test Corp
        </p>
        <p>
          Email: 
          john@test.com
        </p>
        <p>
          Status: 
          Active
        </p>
      </div>
    </div>
  </div>
</div>
`;
